src/message_bus.zig

```diff
@@ -73,60 +191,413 @@ pub const MessageBus = struct {
-        self.allocator.free(message.buffer);
-        self.allocator.destroy(message);
-        self.allocated -= 1;
-    }
+    /// Connection used to send data to another replica
+    /// TODO: This connection will be automatically re-created if it drops,
+    /// using exponential backoff and full jitter.
+    const ReplicaConnection = struct {
+        message_bus: *MessageBus,
+        address: std.net.Address,
+        /// A value of -1 indicates that the connection is not currently active
+        fd: os.socket_t = -1,
+        completion: IO.Completion = undefined,
+        send_queue: FIFO(Envelope) = .{},
+        /// Number of bytes of the current message that have already been sent.
+        bytes_sent: usize = 0,
+
+        fn send_message(self: *ReplicaConnection, envelope: *Envelope) void {
+            const queue_was_empty = self.send_queue.out == null;
+            self.send_queue.push(envelope);
+            // If the queue was not empty, the message will be sent after the
+            // messages currently being sent.
+            if (queue_was_empty) self.send();
+        }
+
+        fn connect(self: *ReplicaConnection) void {
+            assert(self.fd == -1);
+            self.fd = try os.socket(self.address.any.family, os.SOCK_STREAM | os.SOCK_CLOEXEC, 0);
+            self.message_bus.io.connect(
+                *ReplicaConnection,
+                self,
+                complete_connect,
+                &self.completion,
+                self.fd,
+                &self.address.any,
+                self.address.getOsSockLen(),
+            );
+        }
+
+        fn complete_connect(self: *ReplicaConnection, completion: *IO.Completion, result: ConnectError!void) void {
+            result catch |err| {
+                // TODO: maybe don't need to close on *every* error
+                log.err("error connecting to {}: {}", .{ self.address, err });
+                self.close();
+                return;
+            };
+            log.debug("connected to {}", .{self.address});
+            self.send();
+        }
+
+        fn close(self: *ReplicaConnection) void {
+            // Reset bytes_sent to 0 so that we resend the full message on reconnect.
+            self.bytes_sent = 0;
+            self.message_bus.io.close(
+                *ReplicaConnection,
+                self,
+                complete_close,
+                &self.completion,
+                self.fd,
+            );
+        }
+
+        fn complete_close(self: *ReplicaConnection, completion: *IO.Completion, result: ConnectError!void) void {
+            result catch |err| {
+                log.err("error closing connection to replica at {}: {}", .{ self.address, err });
+                return;
+            };
+            log.debug("closed connection to replica at {}", .{self.address});
+            // TODO: add a delay before reconnecting based on exponential backoff/full jitter
+            self.connect();
```

Comment:
We should move the replica reconnect logic outside of the connection `on_close()` callback to decouple the reconnect from the event it's reacting to, which can be safer, especially if anything goes wrong in our exponential backoff logic.

We can then have a `tick()` method on MessageBus which calls `connect_to_replica()` (see my comment in the main conversation of this PR) for all replicas every tick, which would be every 10ms in practice.

Same thing for `maybe_accept()`, and then we can remove the multiple call sites for `maybe_accept()` and replace them with a single call site in `tick()`, which is also safer because it means we no longer need `io.close()` to be asynchronous to avoid cycles when dropping clients within `maybe_accept()` to make space for replicas, and because we no longer have to worry about calling it from all the possible state transitions.

This `tick()` call should be free of overhead since we already need an io_uring timeout to call `Replica.tick()` within the main event loop.

As a bonus it might also come in handy if we ever want to revisit our decision to make MessageBus deterministic. That decision then becomes cheap to revise.

Reply:
Having a `tick()` method sounds like a great way to simplify the logic around `maybe_accept()` and replica connection handling. I'll Implement this while reworking things for full duplex replica connections.

Reply:
The `tick()` approach would then just mean we `accept()` at most every 10ms or so but that's totally fine (100 connections accepted a second would saturate our statically allocated connections pretty quickly) and probably a good control to have.